### PR TITLE
Add ActiveSupport tracer for cache module

### PR DESCRIPTION
**Pull request by frederikspang**
## Description

Implement child spans for ActiveSupport cache according to https://docs.sentry.io/platforms/ruby/tracing/instrumentation/custom-instrumentation/caches-module/

Todo: 

- [x] Specs

### CODE REVIEW

**Review by jespersandnielsen — APPROVED** [review]

**Review by jespersandnielsen — APPROVED** [review]

File: sentry-rails/lib/sentry/rails/tracing/active_support_subscriber.rb
Comment (frederikspang) [reply sentry-rails/lib/sentry/rails/tracing/active_support_subscriber.rb:35]: You're absolutely right. That's a bad leftover from activestorage instrumentation! I'll have a look at this one tomorrow.

Comment (sl0thentr0py) [reply sentry-rails/lib/sentry/rails/tracing/active_support_subscriber.rb:35]: since CI is also failing in many places, I can take this over if you wish but you will need to give me maintainer push access
https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests

Comment (frederikspang) [reply sentry-rails/lib/sentry/rails/tracing/active_support_subscriber.rb]: Be my guest! I'm happy to resolve the errors in here and in #2383 (Also from me/us :) ) 

I've pushed initial changes, but I am not running a full test suite - I'll get that up and running locally as well.

Any way to accept CI runs from this PR automatically? Instead of manual approval each time I do changes :)

Comment (frederikspang) [reply sentry-rails/lib/sentry/rails/tracing/active_support_subscriber.rb]: One thing - Before I go around pure anarchy in terms of `make_basic_app`. With the current set up I am unable to change the Rails.cache store to a memory store, because the block is yielded from an initializer - After initializing the rails app and cache store (Filestore by default).

I can either change the store for all `make_basic_app` setup (Which shouldn't affect much), or we can introduce a new parameter to `make_basic_app(cache_store:)`, but that seems like a bad direction to go.

The issue in specs is, that the Prune method, only works for memorystore.

When that is settled, I'll write specs for all of the other Rails.cache.* methods

Comment (sl0thentr0py) [reply sentry-rails/lib/sentry/rails/tracing/active_support_subscriber.rb]: > Any way to accept CI runs from this PR automatically? Instead of manual approval each time I do changes :)

sadly not, corporate compliance :)

> I can either change the store for all make_basic_app setup (Which shouldn't affect much)

I _think_ nothing else is using/assuming filestore so you can go ahead with changing the default. In case something breaks, we can generalize `make_basic_app.`

Comment (frederikspang) [reply sentry-rails/lib/sentry/rails/tracing/active_support_subscriber.rb]: It's been merged, but for Rails main branch. How do we proceed here, as the "improved" solution is not merged back to "old" Rails versions?

We could just subscribe for the events per definition, and just support the ones, that the users Rails version have?

That wouldn't fix coverage percentage though (Until 8.0 releases, and coverage is run against Rails 8)

Comment (frederikspang) [reply sentry-rails/lib/sentry/rails/tracing/active_support_subscriber.rb]: @sl0thentr0py I have granted you write access for the fork (By transfering the forked repo to my own GitHub instead of work). That should do it, if you had any ideas here!

Comment (sl0thentr0py) [reply sentry-rails/lib/sentry/rails/tracing/active_support_subscriber.rb]: > We could just subscribe for the events per definition, and just support the ones, that the users Rails version have?

Yes, just support what's possible, else just no-op/silently fail.

> That wouldn't fix coverage percentage though (Until 8.0 releases, and coverage is run against Rails 8)

I'll take care of the annoying stuff like coverage, don't worry about it too much. A couple of lines here and there that only run for some versions are ok to be left uncovered since we support such large version matrices and our codecov setup is not taking into account everything.

Comment (frederikspang) [reply sentry-rails/lib/sentry/rails/tracing/active_support_subscriber.rb]: Unless we want to introduce Redis or Memcached in the testing, which is a separate container for the GitHub Actions workflow.

Comment (frederikspang) [reply sentry-rails/lib/sentry/rails/tracing/active_support_subscriber.rb]: @sl0thentr0py Do we need anything else here except internal QA/processing from you guys?
We're looking forward to having this merged soon!

Comment (frederikspang) [reply sentry-rails/lib/sentry/rails/tracing/active_support_subscriber.rb]: No worries!

Comment (frederikspang) [reply sentry-rails/lib/sentry/rails/tracing/active_support_subscriber.rb]: Any update @sl0thentr0py? If you need anything from me, I'm happy to implement to save you guys time.

Comment (solnic) [reply sentry-rails/lib/sentry/rails/tracing/active_support_subscriber.rb]: @frederikspang hey hey, sorry for keeping you waiting, I'm gonna help you wrap this up as Neel is still too busy with other tasks :) I'll test it out locally and do a code review, and in the meantime could you resolve conflicts with the CHANGELOG file? 🙏🏻

Comment (frederikspang) [reply sentry-rails/lib/sentry/rails/tracing/active_support_subscriber.rb]: @solnic Done. Do you want me to squash commits or will you "Squash & Merge" from GitHub?
There may be some broken specs as well - If you allow them to run, I have a chance to take a look at them!

File: sentry-rails/lib/sentry/rails/tracing/active_support_subscriber.rb
Comment (solnic) [inline sentry-rails/lib/sentry/rails/tracing/active_support_subscriber.rb:44]: This has no effect on class method definitions so you could remove it.

**Review by solnic — CHANGES_REQUESTED** [review]
There are some tests failing under older rubies/rails + a couple of rubocop offenses to fix. Oh and conflict with CHANGELOG still has to be resolved.

Other than that, LGTM. I tested it out locally and it seems to work well.

File: sentry-rails/lib/sentry/rails/tracing/active_support_subscriber.rb
Comment (solnic) [reply sentry-rails/lib/sentry/rails/tracing/active_support_subscriber.rb]: > @solnic Done. Do you want me to squash commits or will you "Squash & Merge" from GitHub? There may be some broken specs as well - If you allow them to run, I have a chance to take a look at them!

I'll squash/merge, thanks.

**Comment by frederikspang** [comment]
> There are some tests failing under older rubies/rails + a couple of rubocop offenses to fix. Oh and conflict with CHANGELOG still has to be resolved.
> 
> Other than that, LGTM. I tested it out locally and it seems to work well.

@solnic 
I've adjusted rubocop.
It seems the failing specs are related to Rails 6.0.x release. I just scanned over the changelog for activesupport, and I see nothing related there - Quick test says, `hit` is not in the payload from the AS notification. 

I can drill down into why, but it seems to be out of "our hands".

**Comment by solnic** [comment]
> I've adjusted rubocop.

Awesome :)

> It seems the failing specs are related to Rails 6.0.x release. I just scanned over the changelog for activesupport, and I see nothing related there - Quick test says, `hit` is not in the payload from the AS notification.
> 
> I can drill down into why, but it seems to be out of "our hands".

I think it's perfectly fine to skip it in Rails 6 and add a note to CHANGELOG (and later on in the docs) that it's Rails 7+ only feature.

**Comment by frederikspang** [comment]
> I think it's perfectly fine to skip it in Rails 6 and add a note to CHANGELOG (and later on in the docs) that it's Rails 7+ only feature.

Seems more like a regression in 6.0.x. Works in 5.x and 6.1+
We can skip it, no problem.

**Comment by frederikspang** [comment]
> > I think it's perfectly fine to skip it in Rails 6 and add a note to CHANGELOG (and later on in the docs) that it's Rails 7+ only feature.
> 
> Seems more like a regression in 6.0.x. Works in 5.x and 6.1+ We can skip it, no problem.

I was unable to find anything obvious in Rails 6.0.x releases. I have skipped the spec, and rubocop as well as rspec is passing for me locally. @solnic

**Comment by solnic** [comment]
@frederikspang thanks for addressing remaining issues - could you update your branch to latest master? Coverage should be OK after rebasing as we now merge coverage results from all test runs :)

**Comment by frederikspang** [comment]
> @frederikspang thanks for addressing remaining issues - could you update your branch to latest master? Coverage should be OK after rebasing as we now merge coverage results from all test runs :)

Squashed, and rebased :)

**Comment by frederikspang** [comment]
@solnic I see coverage is kind of misleading in the _spec.rb file.. Rails 8 is not released yet, and is not in the test matrix.
The specs I wrote there are an "improvement" that requires my Rails PR to be release - It's been merged into 8.0, but not backmerged. (Instrumentation of all cache stores).

**Review by solnic — APPROVED** [review]

File: CHANGELOG.md
Comment (solnic) [inline CHANGELOG.md:6]: Could you also mention that there are limitations depending on the Rails version?

Comment (frederikspang) [reply CHANGELOG.md:6]: Done! :)

**Comment by frederikspang** [comment]
Spec coverage should be fixed by https://github.com/getsentry/sentry-ruby/pull/2437 I see :)

**Review by sl0thentr0py — APPROVED** [review]

**Comment by solnic** [comment]
> can I merge this now @solnic?

I merged it :) This PR made me think we could consider adding Rails HEAD to the test matrix - WDYT? Not necessarily as a strict CI step, ie allow it to issue a warning if some specs are failing instead of making CI fail. This way we could catch some issues early on.

**Comment by frederikspang** [comment]
> > can I merge this now @solnic?
> 
> 
> 
> I merged it :) This PR made me think we could consider adding Rails HEAD to the test matrix - WDYT? Not necessarily as a strict CI step, ie allow it to issue a warning if some specs are failing instead of making CI fail. This way we could catch some issues early on.

Great idea! Might throw in Ruby head as well - or just add it "early" for the preview releases

**Comment by st0012** [comment]
> This PR made me think we could consider adding Rails HEAD

We used to do both Rails head and Ruby head. But eventually they just stayed broken most of the time and nobody would check them. More importantly, many times the failures were actually capturing Rails/Ruby/dependencies' bugs, not the SDK's. But we usually only found out after spending hours on it.

So I've adapted the approach to just update the CI matrix and tests when there are release candidates (e.g. #2444).

**Comment by solnic** [comment]
@st0012 yes this makes sense, thanks for the insights!

**Pull request merged**
